mlfinlab/portfolio_optimization/nco.py

```diff
@@ -0,0 +1,362 @@
+                                             Silhouette Coefficient series
+        """
+
+        # Distance matrix from correlation matrix
+        dist_matrix = ((1 - corr.fillna(0)) / 2) ** (1 / 2)
+
+        # Series for Silhouette Coefficients - cluster fit measure
+        silh_coef_optimal = pd.Series()
+
+        # If maximum number of clusters undefined, it's equal to half the number of elements
+        if max_num_clusters is None:
+            max_num_clusters = round(corr.shape[0] / 2)
+
+        # Iterating over the allowed iteration times for k-means
+        for init in range(1, n_init):
+            # Iterating through every number of clusters
+            for num_clusters in range(2, max_num_clusters + 1):
+                # Computing k-means clustering
+                kmeans = KMeans(n_clusters=num_clusters, n_jobs=1, n_init=init)
+                kmeans = kmeans.fit(dist_matrix)
+
+                # Computing a Silhouette Coefficient - cluster fit measure
+                silh_coef = silhouette_samples(dist_matrix, kmeans.labels_)
+
+                # Metrics to compare numbers of clusters
+                stat = (silh_coef.mean() / silh_coef.std(), silh_coef_optimal.mean() / silh_coef_optimal.std())
+
+                # If this is the first metric or better than the previous
+                # we set it as the optimal number of clusters
+                if np.isnan(stat[1]) or stat[0] > stat[1]:
+                    silh_coef_optimal = silh_coef
+                    kmeans_optimal = kmeans
+
+        # Sorting labels of clusters
+        new_index = np.argsort(kmeans_optimal.labels_)
+
+        # Reordering correlation matrix rows
+        corr = corr.iloc[new_index]
+
+        # Reordering correlation matrix columns
+        corr = corr.iloc[:, new_index]
+
+        # Preparing cluster members as dict
+        clusters = {i: corr.columns[np.where(kmeans_optimal.labels_ == i)[0]].tolist() for \
+                    i in np.unique(kmeans_optimal.labels_)}
+
+        # Silhouette Coefficient series
+        silh_coef_optimal = pd.Series(silh_coef_optimal, index=dist_matrix.index)
+
+        return corr, clusters, silh_coef_optimal
+
+    @staticmethod
+    def opt_port(cov, mu_vec=None):
+        """
+        Estimates the Convex Optimization Solution (CVO).
+
+        Uses the covariance matrix and the mu - optimal solution.
+        If mu is the vector of expected values from variables, the result will be
+        a vector of weights with maximum Sharpe ratio.
+        If mu is a vector of ones, the result will be a vector of weights with
+        minimum variance.
+
+        :param cov: (np.array) Covariance matrix of the variables.
+        :param mu_vec: (np.array) Expected value of draws from the variables for maximum Sharpe ratio.
+                              None if outputting the minimum variance portfolio.
+        :return: (np.array) Optimal allocation
```

Comment:
"Optimal allocation" -> "Weights for optimal allocation"

Reply:
Done